server/src/main/java/org/elasticsearch/common/metrics/CounterMetric.java

```diff
@@ -8,29 +8,48 @@
 
 package org.elasticsearch.common.metrics;
 
-import java.util.concurrent.atomic.LongAdder;
+import org.elasticsearch.Assertions;
 
-public class CounterMetric implements Metric {
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 
+public final class CounterMetric {
     private final LongAdder counter = new LongAdder();
+    private final AtomicLong assertingCounter = Assertions.ENABLED ? new AtomicLong() : null;
+
+    private boolean assertNonNegative(long n) {
+        assert n >= 0 : "CounterMetric value must always be non-negative; got: " + n;
+        return true;
+    }
 
     public void inc() {
         counter.increment();
+        assert assertNonNegative(assertingCounter.incrementAndGet());
     }
 
     public void inc(long n) {
         counter.add(n);
+        assert assertNonNegative(assertingCounter.addAndGet(n));
     }
 
     public void dec() {
         counter.decrement();
+        assert assertNonNegative(assertingCounter.decrementAndGet());
     }
 
     public void dec(long n) {
         counter.add(-n);
+        assert assertNonNegative(assertingCounter.addAndGet(-n));
     }
 
     public long count() {
-        return counter.sum();
+        // Retries up to 5 times as the returned value of LongAdder#sum is NOT an atomic snapshot.
```

Comment:
why retry here instead of just directly rounding to 0 in case of negative values?

Reply:
I added a comment in https://github.com/elastic/elasticsearch/pull/71446/commits/cb76a362fbd6fa603d91ad91c0ab8ec6938b797f